packages/form-participation/src/FormControlMixin.ts

```diff
@@ -0,0 +1,517 @@
+import { IElementInternals } from 'element-internals-polyfill';
+import { Constructor, FormControlInterface, IControlHost, Validator } from './types';
+
+const focused = Symbol('focused');
+const forceError = Symbol('forceError');
+const initFormControl = Symbol('initFormControl');
+const onBlur = Symbol('onBlur');
+const onFocus = Symbol('onFocus');
+const onInvalid = Symbol('onInvalid');
+const setValue = Symbol('setValue');
+const shouldShowError = Symbol('shouldShowError');
+const touched = Symbol('touched');
+const validate = Symbol('validate');
+
+export function FormControlMixin<T extends Constructor<HTMLElement & IControlHost>>(SuperClass: T) {
+  class FormControl extends SuperClass {
+    /** Wires up control instances to be form associated */
+    static get formAssociated() {
+      return true;
+    }
+
+    /**
+     * A list of Validator objects that will be evaluated when a control's form
+     * value is modified or optionally when a given attribute changes.
+     *
+     * When a Validator's callback returns false, the entire form control will
+     * be set to an invalid state.
+     */
+    static get formControlValidators(): Validator[] {
+      return [];
+    }
+
+    /**
+     * Allows the FormControl instance to respond to Validator attributes.
+     * For instance, if a given Validator has a `required` attribute, that
+     * validator will be evaluated whenever the host's required attribute
+     * is updated.
+     */
+    static get observedAttributes(): string[] {
+      const validatorAttributes = this.formControlValidators
+        .map(validator => validator.attribute);
+
+      /** @ts-ignore This exits */
+      const observedAttributes = super.observedAttributes || [];
+
+      /** Make sure there are no duplicates inside the attributes list */
+      const attributeSet = new Set([...observedAttributes, ...validatorAttributes])
+      return [...attributeSet];
+    }
+
+    /**
+     * Return the validator associated with a given attribute. If no
+     * Validator is associated with the attribute, it will return null.
+     */
+    static getValidator(attribute: string): Validator {
+      return this.formControlValidators.find(validator =>
+        validator.attribute === attribute
+      );
+    }
+
+    /** The ElementInternals instance for the control. */
+    internals = this.attachInternals() as unknown as IElementInternals;
```

Comment:
Does `attachedInternals` have the same "fire once" requirement as `attachShadow`? What happens if someone had already attached internals in an earlier mixin down the line? _Might_ not be anything we can do, but would be worth documented in that case.

Reply:
Yeah, it's a once-only thing. Let me think on this.